onnx/optimizer/passes/fuse_matmul_add_bias_into_gemm.h

```diff
@@ -0,0 +1,93 @@
+// ATTENTION: The code in this file is highly EXPERIMENTAL.
+// Adventurous users should note that the APIs will probably change.
+
+#pragma once
+
+// Before:
+//   Z = MatMul(X, Y)
+//   B = Z + A
+// After:
+//   B = Gemm(X, Y, A)
+//
+// the pass can handle the case when A is 1D tensor and A.dim[0] == Z.dim[1]
+
+#include <numeric>
+
+#include "onnx/common/assertions.h"
+#include "onnx/optimizer/pass.h"
+
+namespace ONNX_NAMESPACE {
+namespace optimization {
+
+struct FuseMatMulAddBiasIntoGemm final : public PredicateBasedPass {
+  explicit FuseMatMulAddBiasIntoGemm()
+      : PredicateBasedPass(
+            PassType::Fuse,
+            PassEfficiency::Complete,
+            PassOptimizationType::Compute) {}
+  std::string getPassName() const override {
+    return "fuse_matmul_add_bias_into_gemm";
+  }
+  bool patternMatchPredicate(Node* node) override {
+    return node->kind() == kAdd &&
+        node->inputs()[0]->node()->kind() == kMatMul;
+  }
+  bool runTransform(Node* n, Graph& graph, NodeDestroyType& destroy_current)
+      override {
+    // due to current broadcasting's constraint, MatMul has to be the first
+    // operand
+    destroy_current = NodeDestroyType::DestroyZero;
+    auto orig_matmul = n->inputs()[0];
+    auto orig_bias = n->inputs()[1];
+    // check if bias is Const or in graph's initializers
+    if (orig_bias->node()->kind() != kConstant &&
+        orig_bias->node()->kind() != kParam) {
+      return false;
+    }
+    // check if MatMul is only used by Add
+    if (orig_matmul->uses().size() > 1) {
+      return false;
+    }
+    auto bias_shape = orig_bias->sizes();
+    auto weight_shape = orig_matmul->node()->inputs()[1]->sizes();
+    int64_t M = -1;
+    // try to get feature M from weight_shape
+    if (static_cast<int64_t>(weight_shape.size()) == 2 &&
+        weight_shape[1].is_int) {
+      M = weight_shape[1].dim;
+    } else {
+      return false;
+    }
+    // check if bias_shape is compatible
+    int64_t num_el = 1;
+    for (int i = 0; i < static_cast<int64_t>(bias_shape.size()); ++i) {
```

Comment:
Here, to be safe, instead of checking num_el, I would suggest checking len(dim) == 1 or 2, and dim[size-1] == M
Otherwise you will get incorrect output shape.